File: backups/painter_full_2026-07-20/docs/Painter/ph2d_wet_paint/js/engine/tuning.js
// Live tuning registry (spec section 16): one module owning every calibration
// knob the engine reads at runtime. Each def: key, group, EN/PT label, a
// comfortable slider range (free numeric input may exceed it), default, and
// an optional rebuild kind ('brush' = rebuild the bristle texture, 'paper' =
// re-bake the sheet on release, 'render' = repaint).
//
// The `hidden` group holds the gated extensions of spec section 17: they ship
// implemented but default-neutral with no visible sliders. To re-enable one
// in the UI, flip its `hidden` flag - the panel builds itself from this table.

export const KNOB_DEFS = [
  // group, key, en, pt, min, max, step, def, rebuild
  k('paint', 'pigmentPerDab', 'Pigment per dab', 'Pigmento por toque', 0, 2000, 10, 600),
  k('paint', 'paperGate', 'Paper gate', 'Barreira do grao', 0, 1.5, 0.01, 0.6),
  k('paint', 'felt', 'Felt (pores)', 'Feltro (poros)', 0, 0.2, 0.001, 0.01, 'brush'),
  k('paint', 'bristleCount', 'Bristle count', 'Numero de cerdas', 0, 4000, 10, 950, 'brush'),
  k('paint', 'drag', 'Drag', 'Arrasto', 0, 1, 0.01, 0.16),
  k('paint', 'pickup', 'Pickup', 'Captura', 0, 0.2, 0.001, 0.005),
  k('paint', 'intensity', 'Intensity', 'Intensidade', 0, 4, 0.05, 1),
  k('paint', 'bristleStrength', 'Bristle strength', 'Forca das cerdas', 0, 4, 0.05, 1, 'brush'),
  k('paint', 'bristleSize', 'Bristle size', 'Tamanho das cerdas', 0.2, 4, 0.05, 1, 'brush'),
  k('paint', 'spacing', 'Spacing', 'Espacamento', 0.5, 8, 0.5, 2),
  k('paint', 'tipClean', 'Tip clean', 'Limpeza da ponta', 0, 0.05, 0.0005, 0.001),
  k('paint', 'blendForce', 'Blend force', 'Forca do blend', 0, 1, 0.01, 0.8),
  k('paint', 'gateSaturation', 'Gate saturation', 'Saturacao da barreira', 100, 6000, 50, 2000),
  k('water', 'waterPerDab', 'Water per dab', 'Agua por toque', 0, 4, 0.05, 1),
  k('water', 'waterCap', 'Water cap', 'Teto de agua', 1, 30, 0.5, 8),
  k('water', 'evaporation', 'Evaporation', 'Evaporacao', 0, 8, 0.05, 1),
  k('water', 'rewet', 'Re-wet', 'Reumedecimento', 0, 8, 0.05, 1),
  k('water', 'retention', 'Retention', 'Retencao', 0.9, 1, 0.0005, 0.995),
  k('water', 'edgeDarkening', 'Edge darkening', 'Escurecimento de borda', 0, 200, 1, 50),
  k('water', 'baseEvaporation', 'Base evaporation', 'Evaporacao base', 0, 20, 0.1, 2),
  k('physics', 'leveling', 'Leveling', 'Nivelamento', 0, 2, 0.01, 0.5),
  k('physics', 'capillary', 'Capillary', 'Capilaridade', 0, 1, 0.01, 0.2),
  k('physics', 'brake', 'Brake', 'Freio', 0, 4, 0.05, 1.5),
  k('physics', 'gravity', 'Gravity', 'Gravidade', 0, 0.05, 0.0005, 0.005),
  k('physics', 'levelClamp', 'Level clamp', 'Trava de nivelamento', 0, 1, 0.01, 0.2),
  k('physics', 'viscosity', 'Viscosity', 'Viscosidade', 0, 2, 0.01, 0.2),
  k('physics', 'maxVelocity', 'Max velocity', 'Velocidade maxima', 0.2, 3, 0.05, 1),
  k('physics', 'projection', 'Projection', 'Projecao', 0, 1.5, 0.01, 0.7),
  k('physics', 'brakeReach', 'Brake reach', 'Alcance do freio', 1, 12, 0.5, 4),
  k('physics', 'capillaryGate', 'Capillary gate', 'Barreira capilar', 100, 6000, 50, 2000),
  k('tools', 'eraser', 'Eraser', 'Borracha', 0, 1, 0.01, 0.4),
  k('tools', 'dryer', 'Dryer', 'Secador', 0, 1, 0.01, 0.32),
  k('tools', 'blow', 'Blow', 'Sopro', 0, 1, 0.01, 0.2),
  k('tools', 'smear', 'Smear', 'Esfregar', 0, 1, 0.01, 0.8),
  k('paper', 'paperContrast', 'Contrast', 'Contraste', 0.2, 2.5, 0.05, 1, 'paper'),
  k('paper', 'paperFibres', 'Fibres', 'Fibras', 0, 3, 0.05, 1, 'paper'),
  k('paper', 'paperGrooves', 'Grooves', 'Sulcos', 0, 3, 0.05, 1, 'paper'),
  k('paper', 'visualGrain', 'Visual grain', 'Grao visual', 0, 3, 0.05, 1, 'render'),
  k('paper', 'emboss', 'Emboss', 'Relevo', 0, 4, 0.05, 1, 'render'),
  // Render-only master (paper group header): fades the tooth in the
  // APPEARANCE only - the physics never sees it.
  k('paper', 'paperVisibility', 'Paper visibility', 'Visibilidade do papel', 0, 1.5, 0.01, 1, 'render', { master: true }),
  // ---- gated extensions (spec section 17): neutral defaults, no sliders ----
  k('hidden', 'extDiffusion', 'Diffusion', 'Difusao', 0, 0.25, 0.005, 0),
  k('hidden', 'extBackrun', 'Backrun', 'Refluxo (bloom)', 0, 2, 0.05, 0),
  k('hidden', 'extFingering', 'Fingering', 'Dedos de escorrimento', 0, 1, 0.01, 0),
  k('hidden', 'extRivulets', 'Rivulet count', 'Numero de filetes', 4, 60, 1, 15),
  k('hidden', 'extGranulation', 'Physical granulation', 'Granulacao fisica', 0, 1, 0.01, 0),
  k('hidden', 'extStaining', 'Staining', 'Fixacao (staining)', 0, 1, 0.01, 0.5),
  k('hidden', 'extDryBrush', 'Dry brush', 'Pincel seco', 0, 1, 0.01, 0),
  k('hidden', 'extWetSoften', 'Wet-edge softening', 'Suavizacao em molhado', 0, 1, 0.01, 0),
  k('hidden', 'extRakeLight', 'Rake light', 'Luz rasante', 0, 3, 0.05, 0, 'render'),
  k('hidden', 'extValleyGran', 'Valley granulation', 'Granulacao nos vales', 0, 1, 0.01, 0, 'render'),
  k('hidden', 'extWetSheen', 'Wet sheen', 'Brilho umido', 0, 2, 0.05, 0, 'render'),
  k('hidden', 'extGrainDither', 'Grain dither', 'Ruido de grao', 0, 8, 0.1, 0, 'render'),
  k('hidden', 'extEdgeTint', 'Edge tint', 'Tinta de borda', 0, 1, 0.01, 0.5, 'render'),
];

function k(group, key, en, pt, min, max, step, def, rebuild = null, extra = {}) {
  return { group, key, en, pt, min, max, step, def, rebuild, ...extra };
}

export const KNOB_GROUPS = ['paint', 'water', 'physics', 'tools', 'paper'];

/**
 * Create a live tuning instance. `onChange(def, value)` fires on every write
 * (the shell uses it to trigger brush/paper rebuilds and repaints).
 */
export function createTuning(onChange = null) {
  const byKey = new Map(KNOB_DEFS.map((d) => [d.key, d]));
  const values = {};
  for (const d of KNOB_DEFS) values[d.key] = d.def;
  return {
    defs: KNOB_DEFS,
    byKey,
    values,
    get(key) { return values[key]; },
    /** NaN (e.g. from a garbled numeric input) falls back to the default. */
    set(key, v) {
      const d = byKey.get(key);
      if (!d) return;
      const nv = Number.isNaN(v) || typeof v !== 'number' ? d.def : v;
      if (values[key] === nv) return;
      values[key] = nv;
      if (onChange) onChange(d, nv);
    },
    reset(key) { this.set(key, byKey.get(key)?.def); },
    resetGroup(group) {
      for (const d of KNOB_DEFS) if (d.group === group) this.set(d.key, d.def);
    },
    isDefault(key) { return values[key] === byKey.get(key)?.def; },
  };
}

File: crates/ph2d-wet-paint/src/tuning.rs
//! Live tuning registry (port of `tuning.js`, SPEC §16): one module owning
//! every calibration knob the engine reads at runtime.
//!
//! The engine reads knobs through the typed [`Knob`] enum (an array index —
//! cheaper than the JS Map, same semantics); the panel reads the [`KNOB_DEFS`]
//! table for labels/ranges/defaults and writes through [`Tuning::set_by_key`]
//! (NaN falls back to the default, exactly like the JS free numeric input).
//!
//! The `Hidden` group holds the gated extensions of SPEC §17: implemented,
//! default-neutral, no visible sliders.

/// Every calibration knob. Discriminants are indices into [`KNOB_DEFS`] and
/// the value array — the order below IS the table order.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[repr(usize)]
pub enum Knob {
    PigmentPerDab,
    PaperGate,
    Felt,
    BristleCount,
    Drag,
    Pickup,
    Intensity,
    BristleStrength,
    BristleSize,
    Spacing,
    TipClean,
    BlendForce,
    GateSaturation,
    WaterPerDab,
    WaterCap,
    Evaporation,
    Rewet,
    Retention,
    EdgeDarkening,
    BaseEvaporation,
    Leveling,
    Capillary,
    Brake,
    Gravity,
    LevelClamp,
    Viscosity,
    MaxVelocity,
    Projection,
    BrakeReach,
    CapillaryGate,
    Eraser,
    Dryer,
    Blow,
    Smear,
    PaperContrast,
    PaperFibres,
    PaperGrooves,
    VisualGrain,
    Emboss,
    PaperVisibility,
    ExtDiffusion,
    ExtBackrun,
    ExtFingering,
    ExtRivulets,
    ExtGranulation,
    ExtStaining,
    ExtDryBrush,
    ExtWetSoften,
    ExtRakeLight,
    ExtValleyGran,
    ExtWetSheen,
    ExtGrainDither,
    ExtEdgeTint,
}

pub const KNOB_COUNT: usize = 53;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum KnobGroup {
    Paint,
    Water,
    Physics,
    Tools,
    Paper,
    /// SPEC §17 gated extensions: no visible sliders, neutral defaults.
    Hidden,
}

/// What a knob write invalidates (the facade acts on it).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Rebuild {
    /// Rebuild the bristle stamp texture.
    Brush,
    /// Re-bake the paper tile (on slider release).
    Paper,
    /// Appearance only: repaint.
    Render,
}

mod defs;

pub use defs::{KNOB_DEFS, KnobDef};

/// The boot value of every knob, in [`Knob`] discriminant order — `const` so
/// a host's authored-state `DEFAULT` (a `const` item) can carry the exact
/// engine boot (f64 equality gates depend on it).
#[must_use]
pub const fn knob_defaults() -> [f64; KNOB_COUNT] {
    let mut values = [0.0; KNOB_COUNT];
    let mut i = 0;
    while i < KNOB_COUNT {
        values[i] = KNOB_DEFS[i].default;
        i += 1;
    }
    values
}

/// Live knob values. Writes report what they invalidated; the facade reacts
/// (lazy brush rebuild, paper re-bake on release, repaint).
#[derive(Clone)]
pub struct Tuning {
    values: [f64; KNOB_COUNT],
}

impl Default for Tuning {
    fn default() -> Self {
        Tuning {
            values: knob_defaults(),
        }
    }
}

impl Tuning {
    #[inline]
    pub fn get(&self, knob: Knob) -> f64 {
        self.values[knob as usize]
    }

    /// Copy of the whole value array (the per-step params snapshot).
    #[inline]
    pub fn snapshot(&self) -> [f64; KNOB_COUNT] {
        self.values
    }

    pub fn def_of_key(key: &str) -> Option<&'static KnobDef> {
        KNOB_DEFS.iter().find(|d| d.key == key)
    }

    /// Set a knob. NaN (a garbled numeric input) falls back to the default.
    /// Returns the def when the stored value actually changed.
    pub fn set(&mut self, knob: Knob, v: f64) -> Option<&'static KnobDef> {
        let d = &KNOB_DEFS[knob as usize];
        let nv = if v.is_nan() { d.default } else { v };
        if self.values[knob as usize] == nv {
            return None;
        }
        self.values[knob as usize] = nv;
        Some(d)
    }

    /// Panel-side write by registry key. Unknown keys are ignored.
    pub fn set_by_key(&mut self, key: &str, v: f64) -> Option<&'static KnobDef> {
        let d = Self::def_of_key(key)?;
        self.set(d.knob, v)
    }

    pub fn reset(&mut self, knob: Knob) -> Option<&'static KnobDef> {
        self.set(knob, KNOB_DEFS[knob as usize].default)
    }

    /// Reset every knob of a group. Returns the defs that actually changed —
    /// the caller MUST act on their `rebuild` kinds exactly as it would for
    /// single `set` calls (the JS fires onChange per knob; swallowing these
    /// left stale brush textures / unbaked paper — port-verify finding).
    #[must_use = "act on the rebuild kinds of the changed knobs"]
    pub fn reset_group(&mut self, group: KnobGroup) -> Vec<&'static KnobDef> {
        let mut changed = Vec::new();
        for d in KNOB_DEFS.iter() {
            if d.group == group
                && let Some(def) = self.set(d.knob, d.default)
            {
                changed.push(def);
            }
        }
        changed
    }

    pub fn is_default(&self, knob: Knob) -> bool {
        self.values[knob as usize] == KNOB_DEFS[knob as usize].default
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_table_order_is_the_enum_order() {
        for (i, d) in KNOB_DEFS.iter().enumerate() {
            assert_eq!(d.knob as usize, i, "KNOB_DEFS[{i}] holds {:?}", d.knob);
        }
    }

    #[test]
    fn keys_are_unique() {
        for (i, a) in KNOB_DEFS.iter().enumerate() {
            for b in KNOB_DEFS.iter().skip(i + 1) {
                assert_ne!(a.key, b.key);
            }
        }
    }
}
